File: 0149-max-points-on-a-line/0149-max-points-on-a-line.py
class Solution:
    def maxPoints(self, points: List[List[int]]) -> int:
        # 1. For each point determine if it lies on the longest line
        # 2. count all points with same slope
        # 3. Update result with max.
        res =1
        size = len(points)
        for i in range(size):
            p1 = points[i] 
            x1 = p1[0]
            y1 = p1[1]
            count = collections.defaultdict(int)
            for j in range(i+1,size):
                p2 = points[j]
                x2 = p2[0]
                y2 = p2[1]
                delta_x= x2-x1
                delta_y= y2-y1
                if delta_x==0:
                    slope = float("inf")
                else:
                    slope = delta_y/delta_x
                count[slope]+= 1
                res = max(res, count[slope]+ 1)
            
        return res
        

        

File: 0149-max-points-on-a-line/0149-max-points-on-a-line.cpp
#include <algorithm>
#include <iostream>
#include <utility>
#include <vector>
#include <map>
struct cmp {
  /*
    a/b < c/d
    a*b < c*d
  */
  bool operator()(
    const std::pair<int, int> a,
    const std::pair<int, int> b
  ) 
  const {
    if (a.second == 0 && b.second == 0){
      return false;
    }

    if(a.second == 0){
      return true;
    }

    if(b.second == 0){
      return true;
    }

    return a.first * b.second < b.first * a.second;
  }
};
struct Frac {
  int a,b,c,d;
  bool operator < (const Frac& other)const {
    return this ->a* other.b <other.a * this->b;
  }
};

class Solution {
public:
  int maxPoints(std::vector<std::vector<int>> &points) {
    std::vector<std::pair<Frac,int>> v;

    int len = points.size();

    for(int i=0; i<len; ++i){
      int xi = points[i][0];
      int yi = points[i][1];
      for(int j=i+1; j<len; ++j){
        std::pair<int,int> xint;
        int xj = points[j][0];
        int yj = points[j][1];
        int delta_x = xi-xj;
        int delta_y = yi-yj;
        int yintercept;
        if (delta_y == 0) {
          yintercept= yj;
        } else{

        }
        Frac slope = {delta_x,delta_y};
        bool found =0;
        for(std::pair<Frac,int>& y: v){
          Frac z = y.first;
          if(z.a * slope.b == z.b*slope.a && z.c == slope.c){
            found = 1;
            ++y.second;
            break;
          }
        }
        if (!found) v.push_back({slope,1});
      }
    }
    int best = 0;
    for (auto y: v){
      best = std::max(best,y.second);
    }
    int ans = 1;

    while (ans* (ans-1)/2 != best){
      ++ans;
    }
    if (ans== 0){
      ++ans;
    }
    return ans;
  }
};
